File: P1_201902235/factorial.m
clc;
clear;

pkg load database;

conn = pq_connect(setdbopts('dbname', 'parcial1', 'host', 'localhost', 'port', '5432', 'user', 'postgres', 'password', '2008000070*a'));

function [n, resultado] = calcular_factorial()
    while true
        try
            n = input('Ingresa un número entero positivo: ', 's');

            if isempty(regexp(n, '^\d+$', 'once'))
                error('Has ingresado un carácter no numérico. Intenta de nuevo.');
            end

            n = str2double(n);

            if n < 0
                error('El factorial no está definido para números negativos.');
            elseif n == 0 || n == 1
                resultado = 1;
            else
                resultado = prod(2:n);
            end

            % Mostrar el resultado sin notación científica
            fprintf('El factorial de %d es %d\n', n, resultado);
            return;

        catch err
            fprintf('Error: %s\n', err.message);
        end
    end
end

function registrar_factorial(conn, nombre, n, resultado)
    fecha = datestr(now, 'yyyy-mm-dd');
    hora = datestr(now, 'HH:MM:SS');

    sql = 'INSERT INTO factorial (nombre, dato_ingresado, resultado, fecha, hora) VALUES ($1, $2, $3, $4::date, $5::time)';
    pq_exec_params(conn, sql, {nombre, n, resultado, fecha, hora});
end

function guardar_en_txt(nombre, n, resultado, fecha, hora)
    fecha_formateada = datestr(fecha, 'yyyy-mm-dd');
    hora_formateada = datestr(hora, 'HH:MM:SS');

    archivo = fopen('factorial_registros.txt', 'a');
    if archivo == -1
        disp('Error al abrir el archivo de texto.');
        return;
    end

    fprintf(archivo, 'Nombre: %s | Número: %d | Resultado: %d | Fecha: %s | Hora: %s\n', nombre, n, resultado, fecha_formateada, hora_formateada);
    fclose(archivo);
end

function historial_datos(conn)
    sql_query = 'SELECT nombre, dato_ingresado, resultado, fecha, hora FROM factorial;';
    N = pq_exec_params(conn, sql_query);

    if !isempty(N)
        disp('Historial de registros:');
        disp(N);
    else
        disp('No se encontraron registros.');
    end
end

function borrar_datos(conn)
    N = pq_exec_params(conn, 'DELETE FROM factorial;');
    disp('Se han borrado los registros.');
end

while true
    disp('------ Menú ------');
    disp('1. Calcular factorial');
    disp('2. Mostrar historial');
    disp('3. Borrar historial');
    disp('4. Salir');
    opcion = input('Seleccione una opción: ');

    switch opcion
        case 1
            nombre = input('Ingrese su nombre: ', 's');
            [n, resultado] = calcular_factorial();
            registrar_factorial(conn, nombre, n, resultado);
            disp('Resultado registrado en la base de datos.');

            fecha = datestr(now, 'yyyy-mm-dd');
            hora = datestr(now, 'HH:MM:SS');

            guardar_en_txt(nombre, n, resultado, fecha, hora);
            disp('Resultado guardado en el archivo de texto.');

        case 2
            historial_datos(conn);

        case 3
            borrar_datos(conn);

        case 4
            disp('Saliendo del programa...');
            break;

        otherwise
            disp('Opción no válida. Intente nuevamente.');
    endswitch
endwhile

if exist("conn", "var")
    pq_close(conn);
else
    disp("No hay conexión abierta.");
end

File: P1_201902235/factorial.py
import psycopg2
import os
from datetime import datetime
import time
import re

def base_datos():
    try:
        conn = psycopg2.connect(host="localhost", dbname="parcial1", user="postgres", password="2008000070*a", port=5432)
        return conn
    except Exception as e:
        print("Error en la conexión:", e)
        return None

def menu():
    print("\n")
    print("1. Ingresar datos de usuario")
    print("2. Número factorial")
    print("3. Borrado de datos")
    print("4. Salir")

def ingresar_datos_cliente():
    while True:
        print("\nIngrese nombre de usuario:")
        while True:
            nombre = input("Nombre del cliente: ")
            if re.match(r"^[A-Za-z\s]+$", nombre): 
                break
            else:
                print("Error: El nombre solo debe contener letras y espacios. Intenta nuevamente.")
        
        return nombre

def obtener_fecha_hora():
    # Obtener la fecha y hora actual
    now = datetime.now()
    fecha = now.strftime("%Y-%m-%d")  # Solo la fecha
    hora = now.strftime("%H:%M:%S")  # Solo la hora
    return fecha, hora


def borrar():
    try:
        conn = base_datos()
        if conn:
            cursor = conn.cursor()
            cursor.execute("DELETE FROM factorial;")
            conn.commit()
            conn.close()
            print("Todos los datos han sido borrados de la base de datos.")
    except Exception as e:
        print(f"Error al borrar datos: {e}")

def factorial():
    while True:
        try:
            n = input('Ingresa un número entero positivo: ')
            
            if not n.isdigit():
                raise ValueError('Has ingresado un carácter no numérico. Intenta de nuevo.')
            
            n = int(n)
            
            if n < 0:
                raise ValueError('El factorial no está definido para números negativos.')
            elif n == 0 or n == 1:
                resultado = 1
            else:
                resultado = 1
                for i in range(2, n + 1):
                    resultado *= i
            
            # Mostrar el resultado sin notación científica
            print(f'El factorial de {n} es {resultado}')
            return n, str(resultado)
        
        except ValueError as e:
            print(f'Error: {e}')

def principal():
    nombre = None
    while True:
        menu()
        opcion = input("\nSeleccione una opción: ")

        if opcion == "1":
            nombre = ingresar_datos_cliente()

        elif opcion == "2":
            if not nombre:
                print("Primero debe ingresar un nombre de usuario.")
                continue

            numero, resultado = factorial()
            conn = base_datos()
            if conn:
                cursor = conn.cursor()
                fecha, hora = obtener_fecha_hora()  # Obtener la fecha y hora separadas
                cursor.execute("""
                    INSERT INTO factorial (nombre, dato_ingresado, resultado, fecha, hora)
                    VALUES (%s, %s, %s, %s, %s)
                """, (nombre, numero, resultado, fecha, hora))
                conn.commit()
                conn.close()

        elif opcion == "3":
            borrar()

        elif opcion == "4":
            print("Saliendo del sistema...")
            time.sleep(1)
            break

        else:
            print("Opción no válida. Intenta nuevamente.")

principal()
